### PR TITLE
SYNCOPE-1511: Enhance audit queries with additional params 

### PR DESCRIPTION
**Pull request by mmoayyed**
Allows an `AuditQuery` to pass along criteria for events and results, to be employed later on with Syncope's history management features.

### CODE REVIEW

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java [hunk 1]
```diff
@@ -51,9 +52,11 @@ public AuditEntryTO getAuditTO(final AuditEntry auditEntry) {
         }
 
         if (auditEntry.getInput() != null) {
-            auditTO.getInputs().addAll(Arrays.stream(auditEntry.getInput()).
-                    map(input -> ToStringBuilder.reflectionToString(input, ToStringStyle.JSON_STYLE)).
-                    collect(Collectors.toList()));
+            auditTO.getInputs().addAll(Arrays.stream(auditEntry.getInput())
+                .map(input -> input == null
```
Comment (ilgrosso) [inline core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: shouldn't be better to add `filter()` before to skip null input?

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java [hunk 2]
```diff
@@ -16,6 +16,7 @@
  * specific language governing permissions and limitations
  * under the License.
  */
+
```
Comment (ilgrosso) [inline core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: Why this blank line?

File: common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java [hunk 3]
```diff
@@ -20,11 +20,16 @@
...
 
+import java.util.ArrayList;
+import java.util.List;
+
 public class AuditQuery extends AbstractQuery {
 
     private static final long serialVersionUID = -2863334226169614417L;
 
     private String key;
+    private List<String> results = new ArrayList<>();
```
Comment (ilgrosso) [inline common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java]: Better

```
private final List<Result> results = new ArrayList<>();
```

with `org.apache.syncope.common.lib.types.AuditElements.Result`

File: common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java [hunk 4]
```diff
@@ -35,13 +40,41 @@ public void setKey(final String key) {
         this.key = key;
     }
 
+    public List<String> getResults() {
+        return results;
+    }
+
+    @QueryParam("results")
+    public void setResults(final List<String> results) {
+        this.results = results;
```
Comment (ilgrosso) [inline common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java]: safest replacement with 
```
if (results != null) {
  this.results.addAll(results);
}
```

File: common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java [hunk 5]
```diff
@@ -20,11 +20,16 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
 public class AuditQuery extends AbstractQuery {
 
     private static final long serialVersionUID = -2863334226169614417L;
 
     private String key;
+    private List<String> results = new ArrayList<>();
+    private List<String> events = new ArrayList<>();
```
Comment (ilgrosso) [inline common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java]: safer to make also this list `final`

File: common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java [hunk 6]
```diff
@@ -35,13 +40,41 @@ public void setKey(final String key) {
...
+        this.results = results;
+    }
+
+    public List<String> getEvents() {
+        return events;
+    }
+
+    @QueryParam("events")
+    public void setEvents(final List<String> events) {
+        this.events = events;
```
Comment (ilgrosso) [inline common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java]: same as above for `this.results`

File: common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java [hunk 7]
```diff
@@ -35,13 +40,41 @@ public void setKey(final String key) {
...
+    }
+
     public static class Builder extends AbstractQuery.Builder<AuditQuery, Builder> {
 
         public Builder key(final String keyword) {
             getInstance().setKey(keyword);
             return this;
         }
 
+        public Builder results(final List<String> results) {
```
Comment (ilgrosso) [inline common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java]: can you please also add an incremental
```
public Builder result(final String result) {
getInstance().getResults().add(result);
}
```
?

File: common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java [hunk 8]
```diff
@@ -35,13 +40,41 @@ public void setKey(final String key) {
...
+            getInstance().setResults(results);
+            return this;
+        }
+
+        public Builder events(final List<String> events) {
```
Comment (ilgrosso) [inline common/rest-api/src/main/java/org/apache/syncope/common/rest/api/beans/AuditQuery.java]: same as above

File: core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java [hunk 9]
```diff
@@ -54,10 +56,16 @@ private static String buildWhereClauseForEntityKey(final String key) {
             final String key,
             final int page,
             final int itemsPerPage,
-            final List<OrderByClause> orderByClauses) {
-
+            final List<OrderByClause> orderByClauses,
```
Comment (ilgrosso) [inline core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java]: Elsewhere, `orderByClauses` is passed as last argument.

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java (hunk 1 above)
Comment (mmoayyed) [reply core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: Let me confirm, is `null` a valid input in the `getInput()` array? an audit entry might in fact receive a null value as an input parameter?

Comment (ilgrosso) [reply core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: Generally speaking, yes; not for user / group / any object CRUD, of course.

**Pull request merged**